my-app-v/src/hooks/useTodoLogic.jsx

```diff
@@ -0,0 +1,198 @@
+import queryClient from "../utils/queryClient";
+import Swal from 'sweetalert2'
+
+
+export default function useFormLogic() {
+
+    const emptyTodo = {
+        displayName: "",
+        key: "",
+        description: "",
+        attribute: "",
+    }
+
+    const [formData, setFormData] = useState(emptyTodo);
+    const [buttonLabel, setButtonLabel] = useState("Submit");
+    const [editingTodo, setEditingTodo] = useState(null);
+    const [viewTodoItem, setViewTodoItem] = useState(emptyTodo);
+
+    const fireAlertBySwel = (
+        icon = "warning",
+        title = "Alert",
+        text = "Something happened!"
+    ) => {
+        Swal.fire({
+            icon: icon,
+            title: title,
+            text: text,
+            confirmButtonText: "OK"
+        });
+    };
+
+    // Fetch Todos
+    const { data: todos = [], isLoading, error } = useQuery({
+        queryKey: ["todos"],
+        queryFn: fetchTodos,
+    });
+
+    const addTodoMutation = useMutation({
+        mutationFn: addTodo,
+        onSuccess: () => {
+            queryClient.invalidateQueries(["todos"]);
+        },
+    });
+
+    const updateTodoMutation = useMutation({
+        mutationFn: updateTodo,
+        onSuccess: () => {
+            queryClient.invalidateQueries(["todos"]);
+        },
+    });
+
+    const deleteTodoMutation = useMutation({
+        mutationFn: deleteTodo,
+        onSuccess: () => {
+            queryClient.invalidateQueries(["todos"]);
+        },
+    });
+
+    const deleteTodoFun = useCallback((todo) => {
+        // not block delete if that todo item is not being editing.
+        if (editingTodo && editingTodo.id === todo.id) {
+            // make a alert to user that you are editing this todo item
+            //alert("You are editing this todo item. Please cancel editing to delete this item.");
+            fireAlertBySwel("warning", "Editing in Progress!", "You are editing this todo item. Please cancel editing to delete this item.");
+
+            return;
+        }
+
+        // asked user to confirm before deleting
+        Swal.fire({
+            title: "Do you want to delete the todo?",
+            showDenyButton: true,
+            confirmButtonText: "Delete",
+            denyButtonText: `cancel`
+        }).then((result) => {
+            if (result.isConfirmed) {
+                if (viewTodoItem.id === todo.id) {
+                    closeViewTodoComponent();
+                }
+                deleteTodoMutation.mutate({
+                    id: todo.id,
+                    key: todo.key,
+                    displayName: todo.displayName,
+                    description: todo.description,
+                    attribute: todo.attribute,
+                    completed: false,
+                });
+                Swal.fire("Delete!", "", "success");
+            }
+        });
+    }, [deleteTodoMutation]);
+
+    const handleChange = useCallback((e) => {
```

Comment:
[cc] cant we handle form details under TodoForm.
lets say we have useTodoForm which does take care of add and update todo, and validation